File: 2625-increment-submatrices-by-one/2625-increment-submatrices-by-one.go
func rangeAddQueries(n int, queries [][]int) [][]int {
	diff := newMatrix(n + 1)

	for _, q := range queries {
		r1, c1, r2, c2 := q[0], q[1], q[2], q[3]
		diff[r1][c1]++
		diff[r1][c2+1]--
		diff[r2+1][c1]--
		diff[r2+1][c2+1]++
	}

	for i := 0; i < n; i++ {
		for j := 1; j < n; j++ {
			diff[i][j] += diff[i][j-1]
		}
	}

	for i := 1; i < n; i++ {
		for j := 0; j < n; j++ {
			diff[i][j] += diff[i-1][j]
		}
	}

	ans := newMatrix(n)
	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			ans[i][j] = diff[i][j]
		}
	}

	return ans
}

func newMatrix(size int) [][]int {
	res := make([][]int, size)
	for i := 0; i < size; i++ {
		res[i] = make([]int, size)
	}

	return res
}

File: 2625-increment-submatrices-by-one/2625-increment-submatrices-by-one.rs
impl Solution {
    pub fn range_add_queries(n: i32, queries: Vec<Vec<i32>>) -> Vec<Vec<i32>> {
        let n = n as usize;
        let mut diff = vec![vec![0; n + 1]; n + 1];

        for q in queries {
            let (r1, c1, r2, c2) = (q[0] as usize, q[1] as usize, q[2] as usize, q[3] as usize);
            diff[r1][c1] += 1;
            diff[r1][c2 + 1] -= 1;
            diff[r2 + 1][c1] -= 1;
            diff[r2 + 1][c2 + 1] += 1;
        }

        for i in 0..n {
            for j in 1..n {
                diff[i][j] += diff[i][j - 1];
            }
        }

        for i in 1..n {
            for j in 0..n {
                diff[i][j] += diff[i - 1][j];
            }
        }

        let mut ans = vec![vec![0; n]; n];
        for i in 0..n {
            for j in 0..n {
                ans[i][j] = diff[i][j];
            }
        }

        ans
    }
}